scrape/scrape.go

```diff
@@ -595,13 +595,21 @@ func (sp *scrapePool) refreshTargetLimitErr() error {
 }
 
 func (sp *scrapePool) disableEndOfRunStalenessMarkers(targets []*Target) {
+	loops := make([]loop, 0, len(targets))
+
+	// First find all the loops to minimise locking time
 	sp.mtx.Lock()
-	defer sp.mtx.Unlock()
-	for i := range targets {
-		if l, ok := sp.loops[targets[i].hash()]; ok {
-			l.disableEndOfRunStalenessMarkers()
+	for _, t := range targets {
+		if l, ok := sp.loops[t.hash()]; ok {
+			loops = append(loops, l)
 		}
 	}
+	sp.mtx.Unlock()
+
+	// Disable end of run staleness for loops we found.
+	for _, l := range loops {
+		l.disableEndOfRunStalenessMarkers()
```

Comment:
This updates an atomic bool, so it doesn't need locks.

Reply:
Does it take a long time? Otherwise I don't see why we call it out of the for loop.

Reply:
It shouldn't take long time. Maybe it's better to not include this change. The manager mutex was a bigger issue.